test/14.errorHandling.js

```diff
@@ -0,0 +1,412 @@
+  // action = 'D' for deleting the rule, 'I' for inserting the rule
+  // create this rule for both input and output an resolve alterIPTableRules
+  new Promise((res) => {
+    exec(`iptables -${action} INPUT -p tcp -s ${source} -j DROP ;`, (error,stdout,stderr) => {
+      if(stderr)
+        console.error(error)
+      res();
+    })
+  }),
+  new Promise((res) => {
+    exec(`iptables -${action} OUTPUT -p tcp -s ${source} -j DROP ;`, (error,stdout,stderr) => {
+      if(stderr)
+        console.error(error)
+      res();
+    })
+  })
+])
+
+const netDown = async () => await alterIPTableRules('localhost','I');
+const netUp = async () => await alterIPTableRules('localhost','D');
+
+describe("14. Test errors", () => {
+  let driver = null;
+  let conn = null;
+
+  before("open connection, init tables", async () => {
+    await netUp();
+    driver = new Driver();
+
+    conn = await driver.connect(config);
+
+    conn.should.be.ok();
+
+    let err = null;
+
+    try {
+      // create the tables
+      await conn.execute("CREATE TABLE IF NOT EXISTS T1 (F1 int)");
+      await conn.execute("CREATE TABLE IF NOT EXISTS T2 (F1 int)");
+
+      await conn.execute("CREATE TABLE IF NOT EXISTS T3 (F1 int)");
+      await conn.execute("CREATE UNIQUE INDEX UNIQUENESS_INDEX ON T3(F1)");
+
+      // insert data into the tables
+      await conn.execute("INSERT INTO T1 VALUES (1)");
+      await conn.execute("INSERT INTO T2 VALUES (1)");
+
+      const results = await conn.execute("select * from T1");
+      const rows = await results.getRows();
+      await results.close();
+      rows.length.should.not.be.eql(0);
+    } catch (e) {
+      err = e;
+    }
+    should.not.exist(err);
+  });
+
+  after("close connection, delete tables", async () => {
+    let err = null;
+
+    try {
+      // clean up tables
+      await conn.execute("DROP TABLE IF EXISTS T1");
+      await conn.execute("DROP TABLE IF EXISTS T2");
+
+      await conn.execute("DROP TABLE IF EXISTS T3");
+      await conn.execute("DROP INDEX UNIQUENESS_INDEX IF EXISTS");
+
+      // close connections
+      await conn.close();
+    } catch (e) {
+      err = e;
+    }
+    should.not.exist(err);
+  });
+
+  it("14.1 Can detect deadlock", async () => {
+    // open up two connections with autocommit off
+    const c1 = await driver.connect(config);
+    const c2 = await driver.connect(config);
+
+    c1.should.be.ok();
+
+    const update_lock = true;
+    const post_commit = true;
+
+    await c1.execute("set autocommit off");
+    await c2.execute("set autocommit off");
+
+    if (update_lock) {
+      await c2.execute("update T2 set F1=F1+1");
+      await c1.execute("update T1 set F1=F1+1");
+    }
+
+    // now produce a deadlock condition
+    let err = null;
+    try {
+      await Promise.all([
+        c1.execute("update T2 set F1=F1+1"),
+        c2.execute("update T1 set F1=F1+1"),
+      ]);
+    } catch (e) {
+      err = e;
+      console.error(e);
```

Comment:
I think we should be checking for a particular deadlock error just in case we get the wrong error, I would not want the test to pass.